### PR TITLE
Fix call graph with respect to constants.

### LINKED ISSUE #11007: [Sol->Yul] ICE because "No code generated for function even though it is not a constructor."
## Description

```
contract B {
  function g() public {}
}
contract C is B {
  bytes4 constant s2 = B.g.selector;
}
```

throws

https://github.com/ethereum/solidity/blob/2d48052ae50017bc41d8c85f1628b97352e73dcb/libsolidity/codegen/ir/IRGenerator.cpp#L67-L70

when compiled via new code gen `solc --ir test.sol`

Probably related to https://github.com/ethereum/solidity/pull/10973

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #11007

### CODE REVIEW

**Comment by chriseth** [comment]
@cameel can you add tests to this, please?

File: libsolidity/analysis/FunctionCallGraph.cpp [hunk 1]
```diff
@@ -140,7 +141,15 @@ bool FunctionCallGraphBuilder::visit(EmitStatement const& _emitStatement)
 
 bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
 {
-	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	if (auto const* variable = dynamic_cast<VariableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (variable->isConstant())
```
Comment (bshastry) [inline libsolidity/analysis/FunctionCallGraph.cpp:146]: Could be folded in top if `if (x = expr; x->isConstant())`

Comment (bshastry) [reply libsolidity/analysis/FunctionCallGraph.cpp:146]: Perhaps more readable since it is clear that the branch is only for constant variable declarations.

**Comment by cameel** [comment]
Sure.

File: libsolidity/analysis/FunctionCallGraph.cpp [hunk 2]
```diff
@@ -140,7 +141,15 @@ bool FunctionCallGraphBuilder::visit(EmitStatement const& _emitStatement)
...
+		{
+			solAssert(variable->isStateVariable(), "");
```
Comment (bshastry) [inline libsolidity/analysis/FunctionCallGraph.cpp]: Does the following fail the assertion?

```
bytes4 constant x = C.f.selector
contract C { function f() public {} }
```

File: libsolidity/analysis/FunctionCallGraph.cpp (hunk 1 above)
Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp:146]: I personally think that this pattern is less readable. When `expr` is long, the condition ends up at the end and is easy to miss. If I saw that I would likely just see it as checking if `variable` is not null.

File: libsolidity/analysis/FunctionCallGraph.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp]: I think it won't but I'll add this case to the test to make sure.

**Comment by cameel** [comment]
Tests ready. Unfortunately they don't pass so it looks like the fix is not enough. It fixes constants but for immutables, ordinary variables or even referencing the selector in a function body an edge to the internal dispatch is still being added.

@chriseth Should I take it over and fix these cases?

File: libsolidity/analysis/FunctionCallGraph.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp]: Test added. As expected, this case works fine - simply because we're not visiting the file-level constants when building the graph.

**Comment by chriseth** [comment]
@cameel can you summarize the problems? Immutables should have the same behaviour as regular state variables. The expression `A.g.selector` should add an edge from the dispatch to `g`, but not to he dispatch. I think in the IR codegen, we even generated code for `g`

**Comment by cameel** [comment]
Sorry, it's actually an edge from the dispatch, not to the dispatch that is being added.

So anyway, the problems I was referring to were just all of these adding these edges from dispatch:

```solidity
bytes4 immutable pubImmutable = Base.pub.selector;
```
```solidity
bytes4 pubVar = Base.pub.selector;
```
```solidity
return (Base.ext.selector, Base.pub.selector);
```

The functions are not being called here and I think you can't call them internally via the `bytes4` variable that contains the selector? So I thought they should not be in the graph.

But you have a point about the IR codegen. I think it does add them to the dispatch when it's visiting `MemberAccess`. In that case I'll change the test expectations to expect the edge from dispatch.

File: libsolidity/analysis/FunctionCallGraph.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp]: Actually something is triggering this assertion after all, though I don't think it's your example. I see on CircleCI that there are some tests that failed because of it. I'll investigate.

**Comment by cameel** [comment]
@chriseth Expectations updated.

File: libsolidity/analysis/FunctionCallGraph.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp]: Looks like you were right, though your example won't trigger it on its own. It also needs something that references the constant, otherwise it's not visited.

I added an extra check for `variable->isFileLevelVariable()` to the assertion and with this it passes tests for me.

But anyway, good catch!

**Comment by cameel** [comment]
I have also pushed a fixup that fixes the case found by @bshastry in https://github.com/ethereum/solidity/pull/11014#discussion_r582741571

**Comment by chriseth** [comment]
Looks good! Please rebaes.

**Comment by cameel** [comment]
Rebased and squashed.

**Comment by chriseth** [comment]
Approving.

**Review by cameel — APPROVED** [review]
